src/sentry/data_secrecy/api/waive_data_secrecy.py

```diff
@@ -0,0 +1,159 @@
+    ).values_list("ticket_id", flat=True)
+    return [ticket_id for ticket_id in active_zendesk_tickets if ticket_id is not None]
+
+
+@control_silo_endpoint
+class WaiveDataSecrecyEndpoint(ControlSiloOrganizationEndpoint):
+    permission_classes: tuple[type[BasePermission], ...] = (WaiveDataSecrecyPermission,)
+    publish_status = {
+        "GET": ApiPublishStatus.PRIVATE,
+        "PUT": ApiPublishStatus.PRIVATE,
+        "DELETE": ApiPublishStatus.PRIVATE,
+    }
+    owner = ApiOwner.ENTERPRISE
+
+    def get(
+        self,
+        request: Request,
+        organization_context: RpcUserOrganizationContext,
+        organization: RpcOrganization,
+    ) -> Response:
+        """
+        Return current waiver status and a list of all active tickets for an organization.
+        """
+        if not data_access_grant_exists(organization_id=organization.id):
+            return Response(
+                {"detail": "No data secrecy waiver in place."}, status=status.HTTP_404_NOT_FOUND
+            )
+
+        # calling data_access_grant_exists sets the grant status in the cache
+        grant_status = effective_grant_status_cache.get(organization_id=organization.id)
+        if not grant_status.access_start or not grant_status.access_end:
+            logger.error(
+                "EffectiveGrantStatus with valid window missing start or end date",
+                extra={"organization_id": organization.id},
+            )
+            return Response("Effective grant status is malformed", status=status.HTTP_404_NOT_FOUND)
+
+        serialized_grant_status = {
+            "accessStart": grant_status.access_start.isoformat(),
+            "accessEnd": grant_status.access_end.isoformat(),
+            "zendeskTickets": get_active_tickets_for_organization(organization_id=organization.id),
+        }
+        return Response(serialized_grant_status, status=status.HTTP_200_OK)
+
+    def put(
+        self,
+        request: Request,
+        organization_context: RpcUserOrganizationContext,
+        organization: RpcOrganization,
+    ) -> Response:
+        """
+        Manually update or create a data secrecy waiver.
+
+        :param access_end: the timestamp at which data secrecy is reinstated.
+        """
+        validator = DataSecrecyWaiverValidator(data=request.data)
+        if not validator.is_valid():
+            return Response(validator.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        result = validator.validated_data
+
+        DataAccessGrant.objects.update_or_create(
+            organization_id=organization.id,
+            grant_type=DataAccessGrant.GrantType.MANUAL,
+            defaults={
+                "grant_start": timezone.now(),
+                "grant_end": result["access_end"],
+                "granted_by_user_id": request.user.id,
+            },
+        )
+
+        # invalidate cache to force effective grant status recalculation
+        effective_grant_status_cache.delete(organization.id)
```

Comment:
All of these model operations may be better served in some intermediate layer (like a service or repository class). Especially if we want to reuse this logic in other areas, like the Zendesk webhook code later.

We can also use these in testing to remove some boilerplate, like manually updating the cache.